[mlir][linalg][elementwise] Fold broadcast into new elementwise

someoneinjd opened the pull request:
Fold broadcast into new elementwise Op which has affine-map attached.

someoneinjd committed:
[mlir][linalg][elementwise] Fold broadcast into new elementwise

rolfmorel commented on mlir/test/Dialect/Linalg/elementwise/fold.mlir:
```diff
@@ -41,3 +41,47 @@ func.func @binary_transposed(%A : tensor<?x?xf32>, %B: tensor<?x?xf32>, %C: tens
+//
+func.func @unary_broadcasted(%A : tensor<8x32xf32>, %B: tensor<8x16x32xf32>) ->  tensor<8x16x32xf32> {
+  %empty = tensor.empty() : tensor<8x16x32xf32>
+  %broadcasted_A = linalg.broadcast ins(%A : tensor<8x32xf32>) outs(%empty :  tensor<8x16x32xf32>) dimensions = [1]
```
nit: spacing around colon, also elsewhere in file

javedabsar1 commented on mlir/lib/Dialect/Linalg/Transforms/FoldIntoElementwise.cpp:
```diff
@@ -64,6 +64,41 @@ struct FoldTransposePattern : public OpRewritePattern<ElementwiseOp> {
+      AffineMap map = op.getMatchingIndexingMap(operand);
+      auto broadcastOp = operand->get().getDefiningOp<BroadcastOp>();
+
+      if (!map.isIdentity() || !broadcastOp) {
```
Agree with @rengolin  "This becomes an arbitrarily complex producer chain to analyse. We'd need a reasonable algebra here to simplify pairs and propagate transitive and associative properties, reversibility, etc."

 transpose+broadcast + existing map should be seen as projected-permutation. Let me know if you want to go down that path -- I did something similar  here  -- https://github.com/llvm/llvm-project/blame/main/mlir/lib/Dialect/Linalg/Transforms/DecomposeGenericByUnfoldingPermutation.cpp. 
@MaheshRavishankar  had pointed me in this direction originally.

someoneinjd commented:
Thank you all for the feedback.

I have updated the PR based on your suggestions. The pattern now handles the case where the elementwise op has an existing non-identity map (like a transpose).

The implementation now uses `AffineMap::compose` to combine the `elementwise` op's existing indexing map with the map from the `broadcast` / `transpose` op.

I've also added some new test cases that covers this composition logic, similar to the `elementwise(transpose(broadcast(x)))` example that was discussed.

Could you please take another look when you have a chance? I'd appreciate knowing if this implementation is on the right track.

jopperm reviewed (commented):
The approach looks good to me. I want to point out that this already goes a bit beyond @rengolin's option 1: as the pass applies the patterns greedily, it handles chains of transposes and broadcasts automatically. IMHO this is useful.

I wonder whether the pattern should only fold if the elementwise's operand's map satisfies `isProjectedPermutation()`. My current understanding is that composing with the maps corresponding to the `linalg.transpose` and `linalg.broadcast` ops would preserve that property, and should help reversibility via the transform that @javedabsar1 linked.

someoneinjd commented:
Ping @jopperm @javedabsar1

rolfmorel reviewed (approved):
Great work -- looks really clean and makes good use of the GreedyRewriter to keep on folding up to a fixed point (with arbitrary long chains of transposes and broadcasts on an elementwise's operand)! 🙂 

I will second @jopperm and @javedabsar1's call for restricting the folding to projected permutations (i.e. needing the map for the operand of elementwise to already be a projected permutation), at least for now. We can expand this later if there's a need.

EDIT: please also update the (updated) description of `-linalg-fold-into-elementwise`.

For controllability, the projected permutation restriction might help as then we could always unfold again via DecomposeGenericByUnfoldingPermutation. Whether the controllability should be on this PR's patterns or DecomposeGenericByUnfoldingPermutation is something a follow-up PR could address. That is, having `-linalg-fold-into-elementwise` be fully greedy for now is fine by me.

On the topic of follow-ups: couldn't the same folding be applied to elementwise's result?

In any case, please wait for @jopperm and either of @javedabsar1 or @rengolin to approve before merging.

someoneinjd commented:
Thank you for the feedback. I have updated the PR with the following changes:

* Restricted folding to cases where the consumer indexing map is a projected permutation to ensure controllability and reversibility.
* Updated the pass description in Passes.td to include linalg.broadcast.

Please let me know if there are any further adjustments needed. Ready for another look.

jopperm commented on mlir/include/mlir/Dialect/Linalg/Passes.td:
```diff
@@ -167,12 +167,12 @@ def LinalgInlineScalarOperandsPass : Pass<"linalg-inline-scalar-operands"> {
-    Fold transpose ops that feed `linalg.elementwise` into the elementwise op
-    by updating its indexing maps. `linalg.transpose` producers whose consumer
+    Fold transpose and broadcast ops that feed `linalg.elementwise` into the elementwise op
+    by updating its indexing maps. `linalg.transpose` and `linalg.broadcast` producers whose consumer
```
Nit: The rest of the file seems to adhere to 80 cols in the description text.

jopperm reviewed (approved):
Thanks for iterating! Please consider adding one test with a non-permuted-projection affine map. Otherwise this looks great now.

rolfmorel commented:
I tried to ping @javedabsar1 on Discord last week. No response since. His last response on this PR was back in January.

@rengolin, @jopperm, how about overriding the block and letting @someoneinjd get this in? IMO, they have addressed everything we brought up, including Javed's comments, and persevered all this time (over 5 months now). The patch is a worthwhile addition, IMHO.

javedabsar1 reviewed (approved):
Just some suggested variable name and comment suggestions to help improve understandability of code. 
Otherwise LGTM, so just make the changes and land.
